ding/policy/bc.py

```diff
@@ -0,0 +1,196 @@
+import math
+import torch
+import torch.nn as nn
+import copy
+from torch.optim import SGD
+from torch.optim.lr_scheduler import LambdaLR
+from typing import List, Dict, Any, Tuple, Union, Optional
+from collections import namedtuple, deque
+from easydict import EasyDict
+from ding.policy import Policy
+from ding.model import model_wrap
+from ding.torch_utils import to_device
+from ding.utils import EasyTimer
+from ding.utils.data import default_collate, default_decollate
+from ding.rl_utils import q_nstep_td_data, q_nstep_sql_td_error, get_nstep_return_data, get_train_sample
+from ding.worker.collector.interaction_serial_evaluator import InteractionSerialEvaluator
+from ding.utils import POLICY_REGISTRY
+
+
+@POLICY_REGISTRY.register('bc')
+class BehaviourCloningPolicy(Policy):
+    config = dict(
+        type='bc',
+        on_policy=False,
+        collect=dict(
+            # (int) Only one of [n_sample, n_step, n_episode] shoule be set
+            #n_sample=8,  # collect 8 samples and put them in collector
+            # (int) Cut trajectories into pieces with length "unroll_len".
+            unroll_len=1,
+        ),
+    )
+
+    def _init_learn(self):
+        self._optimizer = SGD(
+            self._model.parameters(),
+            lr=self._cfg.learn.learning_rate,
+            weight_decay=self._cfg.learn.weight_decay,
+            momentum=0.9
+        )
+        self._timer = EasyTimer(cuda=True)
+
+        def lr_scheduler_fn(epoch):
+            if epoch <= self._cfg.learn.warmup_epoch:
+                return self._cfg.learn.warmup_lr / self._cfg.learn.learning_rate
+            else:
+                ratio = epoch // self._cfg.learn.decay_epoch
+                return math.pow(self._cfg.learn.decay_rate, ratio)
+
+        self._lr_scheduler = LambdaLR(self._optimizer, lr_scheduler_fn)
+        self._lr_scheduler.step()
+        self._learn_model = model_wrap(self._model, 'base')
+        self._learn_model.reset()
+
+        self._ce_loss = nn.CrossEntropyLoss()
+
+    def _forward_learn(self, data):
+        if self._cuda:
+            data = to_device(data, self._device)
+        self._learn_model.train()
+
+        with self._timer:
+            if self.cfg.eval.evaluator.cfg_type == 'MetricSerialEvaluatorDict':
```

Comment:
Why use this branch in `forward_learn` method, data pre-processing will be executed before here.